src/probnum/randprocs/kernels/_matern.py

```diff
@@ -65,53 +90,135 @@ class Matern(Kernel, IsotropicMixin):
-            raise ValueError(f"Lengthscale l={self.lengthscale} must be positive.")
-        self.nu = _utils.as_numpy_scalar(nu)
-        if not self.nu > 0:
-            raise ValueError(f"Hyperparameter nu={self.nu} must be positive.")
-
         super().__init__(input_shape=input_shape)
 
-    def _evaluate(self, x0: np.ndarray, x1: Optional[np.ndarray] = None) -> np.ndarray:
-        distances = self._euclidean_distances(x0, x1)
-
-        # Kernel matrix computation dependent on differentiability
-        if self.nu == 0.5:
-            return np.exp(-1.0 / self.lengthscale * distances)
-
-        if self.nu == 1.5:
-            scaled_distances = np.sqrt(3) / self.lengthscale * distances
-            return (1.0 + scaled_distances) * np.exp(-scaled_distances)
-
-        if self.nu == 2.5:
-            scaled_distances = np.sqrt(5) / self.lengthscale * distances
-            return (1.0 + scaled_distances + scaled_distances**2 / 3.0) * np.exp(
-                -scaled_distances
-            )
-        if self.nu == 3.5:
-            scaled_distances = np.sqrt(7) / self.lengthscale * distances
-            # Using Horner's method speeds up computations substantially
-            return (
-                1.0
-                + (1.0 + (2.0 / 5.0 + scaled_distances / 15.0) * scaled_distances)
-                * scaled_distances
-            ) * np.exp(-scaled_distances)
-
-        if self.nu == np.inf:
-            return np.exp(-1.0 / (2.0 * self.lengthscale**2) * distances**2)
-
-        # The modified Bessel function K_nu is not defined for z=0
-        distances = np.maximum(distances, np.finfo(distances.dtype).eps)
-
-        scaled_distances = np.sqrt(2 * self.nu) / self.lengthscale * distances
-        return (
-            2 ** (1.0 - self.nu)
-            / scipy.special.gamma(self.nu)
-            * scaled_distances**self.nu
-            * scipy.special.kv(self.nu, scaled_distances)
+        # Smoothness parameter ν
+        nu = _utils.as_numpy_scalar(nu)
+
+        if nu <= 0:
+            raise ValueError(f"Hyperparameter nu={nu} must be positive.")
+
+        self._nu = nu
+
+        # Input lengthscales
+        lengthscales = np.asarray(
+            lengthscales if lengthscales is not None else 1.0,
+            dtype=np.double,
+        )
+
+        if np.any(lengthscales <= 0):
+            raise ValueError(f"All lengthscales l={lengthscales} must be positive.")
+
+        np.broadcast_to(  # Check if the lengthscales broadcast to the input dimension
+            lengthscales, self.input_shape
+        )
+
+        self._lengthscales = lengthscales
+
+    @property
+    def nu(self) -> ScalarType:
+        r"""Smoothness parameter :math:`\nu`."""
+        return self._nu
+
+    @functools.cached_property
+    def p(self) -> Optional[int]:
+        r"""Order :math:`p` of the polynomial part of the kernel if :math:`\nu` is a
+        half integer :math:`\nu = p + \frac{1}{2}`. Otherwise this is set to `None`."""
```

Comment:
```suggestion
        r"""Order :math:`p` of the polynomial part of the kernel, if :math:`\nu` is a
        half integer, i.e. :math:`\nu = p + \frac{1}{2}`. In this case :math:`p` corresponds
        to the differentiability of the sample paths of a GP with the given kernel.
        Otherwise this is set to `None`."""
```